### DISCUSSION TITLE
MOG2 vs CNT

### DISCUSSION THREAD

**Discussion by ceewanna** [Breakthrough/DVR-Scan#115] [General]
After testing both methods, it seems CNT always produces at least 1 video event for every input file. This happens regardless what threshold or kernel is set. My test set both CNT and MOG2 gets triggered at the same value of threshold. 

I left the CNT on one night (which usually should produces almost no motion video) but there was every motion video for every input file. Then I tested this specifically to a clean input video and found it produces redundant output as outlined above. 

I have no idea whether this is the nature of CNT or not.

**Comment by Breakthrough** [comment]
Hi, do these always occur at the same point in time?  E.g. is it always at the beginning of the input file?  If you could provide some output logs or samples, that would be very helpful to finding the issue.

Thanks!

**Reply by ceewanna** [reply]
**### testdvr.sh**
#!/bin/bash

dname=$(dirname $1)

#rm testset/temp/$fname.DSME*.mp4

mkdir -p $dname/res

#for f in testset/ch5/*
for f in $1
do
        fname=$(basename $f .mp4)
        rm $dname/res/$fname*
        dvr-scan -i $f -c testdvr.cfg -d $dname/res
        ls -s $dname/res/$fname*
done

**### testset.sh**
#!/bin/bash
./testdvr.sh testset/ch2-front-20221005-175701.mp4
./testdvr.sh testset/ch2-front-20221005-180001.mp4
./testdvr.sh testset/ch5-pavil-20221005-175701.mp4
./testdvr.sh testset/ch5-pavil-20221007-000001.mp4

**### testdvr.cfg**
bg-subtractor = CNT
output-mode = ffmpeg

ffmpeg-input-args = -v error
ffmpeg-output-args = -preset ultrafast -crf 32 -c:v libx265 -c:a copy
#ffmpeg-output-args = -preset ultrafast -crf 32 -c:v copy -c:a copy

min-event-length = 0.2s
threshold = 0.10
kernel-size = 15

time-before-event = 2s
time-post-event = 8s

downscale-factor = 3
frame-skip = 1


### **Output**
[DVR-Scan] DVR-Scan v1.5.1
[DVR-Scan] Loading config from file:
  testdvr.cfg
[DVR-Scan] Initializing scan context...
[DVR-Scan] Opened video ch2-front-20221005-175701.mp4 (2560 x 1440 at 24.754 FPS).
[DVR-Scan] Scanning input video for motion events...
[DVR-Scan] Splitting events using ffmpeg, first event:
[DVR-Scan] ffmpeg -y -nostdin -v error -ss 00:00:00.040 -i testset/ch2-front-20221005-175701.mp4 -t 00:00:09.211 -preset ultrafast -crf 32 -c:v libx265 -c:a copy testset/res/ch2-front-20221005-175701.DSME_0001.mp4
[DVR-Scan] ffmpeg output:

x265 [info]: HEVC encoder version 3.4
x265 [info]: build info [Linux][GCC 9.3.0][64 bit] 8bit+10bit+12bit
x265 [info]: using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x265 [info]: Main profile, Level-5 (Main tier)
x265 [info]: Thread pool created using 12 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 3 / wpp(45 rows)
x265 [info]: Coding QT: max CU size, min CU size : 32 / 16
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : dia / 57 / 0 / 2
x265 [info]: Lookahead / bframes / badapt        : 5 / 3 / 0
x265 [info]: b-pyramid / weightp / weightb       : 1 / 0 / 0
x265 [info]: References / ref-limit  cu / depth  : 1 / off / off
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 1 / 0.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-32.0 / 0.60
x265 [info]: tools: rd=2 psy-rd=2.00 early-skip rskip mode=1 tmvp fast-intra
x265 [info]: tools: strong-intra-smoothing lslices=8 deblock
x265 [info]: frame I:      1, Avg QP:33.02  kb/s: 52278.20
x265 [info]: frame P:     58, Avg QP:35.77  kb/s: 2286.19 
x265 [info]: frame B:    172, Avg QP:39.67  kb/s: 46.95   
x265 [info]: consecutive B-frames: 1.7% 1.7% 0.0% 96.6% 

encoded 231 frames in 9.24s (24.99 fps), 835.30 kb/s, Avg QP:38.66

[DVR-Scan] ffmpeg output:

x265 [info]: HEVC encoder version 3.4
x265 [info]: build info [Linux][GCC 9.3.0][64 bit] 8bit+10bit+12bit
x265 [info]: using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x265 [info]: Main profile, Level-5 (Main tier)
x265 [info]: Thread pool created using 12 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 3 / wpp(45 rows)
x265 [info]: Coding QT: max CU size, min CU size : 32 / 16
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : dia / 57 / 0 / 2
x265 [info]: Lookahead / bframes / badapt        : 5 / 3 / 0
x265 [info]: b-pyramid / weightp / weightb       : 1 / 0 / 0
x265 [info]: References / ref-limit  cu / depth  : 1 / off / off
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 1 / 0.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-32.0 / 0.60
x265 [info]: tools: rd=2 psy-rd=2.00 early-skip rskip mode=1 tmvp fast-intra
x265 [info]: tools: strong-intra-smoothing lslices=8 deblock
x265 [info]: frame I:      2, Avg QP:32.52  kb/s: 84143.10
x265 [info]: frame P:     71, Avg QP:36.16  kb/s: 1276.95 
x265 [info]: frame B:    213, Avg QP:39.67  kb/s: 54.78   
x265 [info]: consecutive B-frames: 1.4% 1.4% 1.4% 95.9% 

encoded 286 frames in 6.68s (42.78 fps), 946.22 kb/s, Avg QP:38.75

[DVR-Scan] Processed 1510 frames read in 18.1 secs (avg 83.5 FPS).
[DVR-Scan] Detected 2 motion events in input.
[DVR-Scan] List of motion events:
-------------------------------------------------------------
|   Event #    |  Start Time  |   Duration   |   End Time   |
-------------------------------------------------------------
|  Event    1  |  00:00:00.0  |  00:00:09.2  |  00:00:09.3  |
|  Event    2  |  00:00:09.9  |  00:00:11.4  |  00:00:21.3  |
-------------------------------------------------------------
[DVR-Scan] Comma-separated timecode values:
00:00:00.040,00:00:09.251,00:00:09.857,00:00:21.289
[DVR-Scan] Motion events written to disk.
 992 testset/res/ch2-front-20221005-175701.DSME_0001.mp4
1388 testset/res/ch2-front-20221005-175701.DSME_0002.mp4
[DVR-Scan] DVR-Scan v1.5.1
[DVR-Scan] Loading config from file:
  testdvr.cfg
[DVR-Scan] Initializing scan context...
[DVR-Scan] Opened video ch2-front-20221005-180001.mp4 (2560 x 1440 at 24.738 FPS).
[DVR-Scan] Scanning input video for motion events...
[DVR-Scan] Splitting events using ffmpeg, first event:
[DVR-Scan] ffmpeg -y -nostdin -v error -ss 00:00:00.040 -i testset/ch2-front-20221005-180001.mp4 -t 00:00:09.217 -preset ultrafast -crf 32 -c:v libx265 -c:a copy testset/res/ch2-front-20221005-180001.DSME_0001.mp4
[DVR-Scan] ffmpeg output:

x265 [info]: HEVC encoder version 3.4
x265 [info]: build info [Linux][GCC 9.3.0][64 bit] 8bit+10bit+12bit
x265 [info]: using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x265 [info]: Main profile, Level-5 (Main tier)
x265 [info]: Thread pool created using 12 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 3 / wpp(45 rows)
x265 [info]: Coding QT: max CU size, min CU size : 32 / 16
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : dia / 57 / 0 / 2
x265 [info]: Lookahead / bframes / badapt        : 5 / 3 / 0
x265 [info]: b-pyramid / weightp / weightb       : 1 / 0 / 0
x265 [info]: References / ref-limit  cu / depth  : 1 / off / off
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 1 / 0.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-32.0 / 0.60
x265 [info]: tools: rd=2 psy-rd=2.00 early-skip rskip mode=1 tmvp fast-intra
x265 [info]: tools: strong-intra-smoothing lslices=8 deblock
x265 [info]: frame I:      1, Avg QP:33.02  kb/s: 76828.00
x265 [info]: frame P:     58, Avg QP:35.81  kb/s: 1713.05 
x265 [info]: frame B:    172, Avg QP:39.67  kb/s: 47.37   
x265 [info]: consecutive B-frames: 1.7% 1.7% 0.0% 96.6% 

encoded 231 frames in 8.79s (26.28 fps), 797.97 kb/s, Avg QP:38.67

[DVR-Scan] Processed 1508 frames read in 12.1 secs (avg 124.8 FPS).
[DVR-Scan] Detected 1 motion events in input.
[DVR-Scan] List of motion events:
-------------------------------------------------------------
|   Event #    |  Start Time  |   Duration   |   End Time   |
-------------------------------------------------------------
|  Event    1  |  00:00:00.0  |  00:00:09.2  |  00:00:09.3  |
-------------------------------------------------------------
[DVR-Scan] Comma-separated timecode values:
00:00:00.040,00:00:09.257
[DVR-Scan] Motion events written to disk.
948 testset/res/ch2-front-20221005-180001.DSME_0001.mp4
[DVR-Scan] DVR-Scan v1.5.1
[DVR-Scan] Loading config from file:
  testdvr.cfg
[DVR-Scan] Initializing scan context...
[DVR-Scan] Opened video ch5-pavil-20221005-175701.mp4 (1920 x 1080 at 25.033 FPS).
[DVR-Scan] Scanning input video for motion events...
[DVR-Scan] Splitting events using ffmpeg, first event:
[DVR-Scan] ffmpeg -y -nostdin -v error -ss 00:00:00.040 -i testset/ch5-pavil-20221005-175701.mp4 -t 00:00:09.188 -preset ultrafast -crf 32 -c:v libx265 -c:a copy testset/res/ch5-pavil-20221005-175701.DSME_0001.mp4
[DVR-Scan] ffmpeg output:

x265 [info]: HEVC encoder version 3.4
x265 [info]: build info [Linux][GCC 9.3.0][64 bit] 8bit+10bit+12bit
x265 [info]: using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x265 [info]: Main profile, Level-4 (Main tier)
x265 [info]: Thread pool created using 12 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 3 / wpp(34 rows)
x265 [info]: Coding QT: max CU size, min CU size : 32 / 16
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : dia / 57 / 0 / 2
x265 [info]: Lookahead / bframes / badapt        : 5 / 3 / 0
x265 [info]: b-pyramid / weightp / weightb       : 1 / 0 / 0
x265 [info]: References / ref-limit  cu / depth  : 1 / off / off
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 1 / 0.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-32.0 / 0.60
x265 [info]: tools: rd=2 psy-rd=2.00 early-skip rskip mode=1 tmvp fast-intra
x265 [info]: tools: strong-intra-smoothing lslices=6 deblock
x265 [info]: frame I:      1, Avg QP:33.03  kb/s: 56780.20
x265 [info]: frame P:     58, Avg QP:36.61  kb/s: 897.54  
x265 [info]: frame B:    171, Avg QP:39.67  kb/s: 27.03   
x265 [info]: consecutive B-frames: 3.4% 0.0% 0.0% 96.6% 

encoded 230 frames in 3.29s (69.85 fps), 493.31 kb/s, Avg QP:38.87

[DVR-Scan] ffmpeg output:

x265 [info]: HEVC encoder version 3.4
x265 [info]: build info [Linux][GCC 9.3.0][64 bit] 8bit+10bit+12bit
x265 [info]: using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x265 [info]: Main profile, Level-4 (Main tier)
x265 [info]: Thread pool created using 12 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 3 / wpp(34 rows)
x265 [info]: Coding QT: max CU size, min CU size : 32 / 16
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : dia / 57 / 0 / 2
x265 [info]: Lookahead / bframes / badapt        : 5 / 3 / 0
x265 [info]: b-pyramid / weightp / weightb       : 1 / 0 / 0
x265 [info]: References / ref-limit  cu / depth  : 1 / off / off
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 1 / 0.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-32.0 / 0.60
x265 [info]: tools: rd=2 psy-rd=2.00 early-skip rskip mode=1 tmvp fast-intra
x265 [info]: tools: strong-intra-smoothing lslices=6 deblock
x265 [info]: frame I:      2, Avg QP:32.58  kb/s: 50018.30
x265 [info]: frame P:    122, Avg QP:36.44  kb/s: 882.41  
x265 [info]: frame B:    364, Avg QP:39.67  kb/s: 52.62   
x265 [info]: consecutive B-frames: 1.6% 0.8% 0.0% 97.6% 

encoded 488 frames in 3.61s (135.36 fps), 464.85 kb/s, Avg QP:38.83

[DVR-Scan] Processed 1526 frames read in 8.3 secs (avg 183.0 FPS).
[DVR-Scan] Detected 2 motion events in input.
[DVR-Scan] List of motion events:
-------------------------------------------------------------
|   Event #    |  Start Time  |   Duration   |   End Time   |
-------------------------------------------------------------
|  Event    1  |  00:00:00.0  |  00:00:09.2  |  00:00:09.2  |
|  Event    2  |  00:00:26.5  |  00:00:19.5  |  00:00:46.1  |
-------------------------------------------------------------
[DVR-Scan] Comma-separated timecode values:
00:00:00.040,00:00:09.228,00:00:26.525,00:00:46.060
[DVR-Scan] Motion events written to disk.
 564 testset/res/ch5-pavil-20221005-175701.DSME_0001.mp4
1120 testset/res/ch5-pavil-20221005-175701.DSME_0002.mp4
[DVR-Scan] DVR-Scan v1.5.1
[DVR-Scan] Loading config from file:
  testdvr.cfg
[DVR-Scan] Initializing scan context...
[DVR-Scan] Opened video ch5-pavil-20221007-000001.mp4 (1920 x 1080 at 25.033 FPS).
[DVR-Scan] Scanning input video for motion events...
[DVR-Scan] Splitting events using ffmpeg, first event:
[DVR-Scan] ffmpeg -y -nostdin -v error -ss 00:00:00.040 -i testset/ch5-pavil-20221007-000001.mp4 -t 00:00:09.188 -preset ultrafast -crf 32 -c:v libx265 -c:a copy testset/res/ch5-pavil-20221007-000001.DSME_0001.mp4
[DVR-Scan] ffmpeg output:

x265 [info]: HEVC encoder version 3.4
x265 [info]: build info [Linux][GCC 9.3.0][64 bit] 8bit+10bit+12bit
x265 [info]: using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x265 [info]: Main profile, Level-4 (Main tier)
x265 [info]: Thread pool created using 12 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 3 / wpp(34 rows)
x265 [info]: Coding QT: max CU size, min CU size : 32 / 16
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : dia / 57 / 0 / 2
x265 [info]: Lookahead / bframes / badapt        : 5 / 3 / 0
x265 [info]: b-pyramid / weightp / weightb       : 1 / 0 / 0
x265 [info]: References / ref-limit  cu / depth  : 1 / off / off
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 1 / 0.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-32.0 / 0.60
x265 [info]: tools: rd=2 psy-rd=2.00 early-skip rskip mode=1 tmvp fast-intra
x265 [info]: tools: strong-intra-smoothing lslices=6 deblock
x265 [info]: frame I:      1, Avg QP:33.79  kb/s: 6068.60 
x265 [info]: frame P:     58, Avg QP:36.81  kb/s: 45.28   
x265 [info]: frame B:    171, Avg QP:39.67  kb/s: 25.16   
x265 [info]: consecutive B-frames: 3.4% 0.0% 0.0% 96.6% 

encoded 230 frames in 3.01s (76.34 fps), 56.51 kb/s, Avg QP:38.92

[DVR-Scan] Processed 1526 frames read in 4.6 secs (avg 334.7 FPS).
[DVR-Scan] Detected 1 motion events in input.
[DVR-Scan] List of motion events:
-------------------------------------------------------------
|   Event #    |  Start Time  |   Duration   |   End Time   |
-------------------------------------------------------------
|  Event    1  |  00:00:00.0  |  00:00:09.2  |  00:00:09.2  |
-------------------------------------------------------------
[DVR-Scan] Comma-separated timecode values:
00:00:00.040,00:00:09.228
[DVR-Scan] Motion events written to disk.
72 testset/res/ch5-pavil-20221007-000001.DSME_0001.mp4

**Reply by ceewanna** [reply]
For every file processed, it always gives a free event at the beginning. I once saw it was the first 3.x seconds but this test batch it is 9 something.

**Reply by Breakthrough** [reply]
I will look into this some more, thank you.  It would also help to see the output of `-mo mask.avi` for the first few seconds of videos where this happens - would you be willing to share that / see if anything interesting happens there?  Thanks!

**Reply by Breakthrough** [reply]
I think I see the issue, indeed this is a bug.  Thanks for letting me know, will turn this into a bug report and have it fixed in the next release.

**Reply by ceewanna** [reply]
I just tested it with -mo on one of my files. I saw one square white mask once at the beginning and the the rest masks seem to follow the path of the moving object in my video.

**Reply by ceewanna** [reply]
On another file, where there is only 1 event (9.2 sec) detected, I saw a long period of black screen but no one white square happened. The long period of black screen at the beginning seems to be similar to the one I mentioned above. So I guess the one brink at the beginning may be related to the real event in my video. Not so sure if all these help you identify the problem. Let me know if you need anything more.

**Reply by Breakthrough** [reply]
I filed #119 to follow up on this in the future.  I see this occasionally too, so may have to start ignoring the first frame output of CNT to compensate.  Thanks for the report!
